low-level/files/opt/seagate/sspl/setup/resource_map/server.py

```diff
@@ -32,18 +42,365 @@ class ServerMap(ResourceMap):
+
+        # Retrieve the by-id symlink for each drive and save in a dict with the drive path as key
+        for block_dev in block_devs:
+            try:
+                if self._disk_objects[block_dev['path']].get('org.freedesktop.UDisks2.Block') is not None:
+                    # Obtain the list of symlinks for the block device
+                    udisk_block = self._disk_objects[block_dev['path']]["org.freedesktop.UDisks2.Block"]
+                    symlinks = self._sanitize_dbus_value(udisk_block["Symlinks"])
+
+                    # Parse out the wwn symlink if it exists otherwise use the by-id
+                    for symlink in symlinks:
+                        if "wwwn" in symlink:
+                            self._drive_by_id[udisk_block["Drive"]] = symlink
+                        elif "by-id" in symlink:
+                            self._drive_by_id[udisk_block["Drive"]] = symlink
+                        # TODO:  Improve logic for getting resource_id
+                        # Current approch for getting resource_id is to check "phy" in by-path
+                        # symlink. If "phy" is in by-path use path of that drive for resource_id
+                        elif "by-path" in symlink:
+                            if "phy" in symlink:
+                                self._drive_by_path[udisk_block["Drive"]] = symlink[len("/dev/disk/by-path/"):]
+
+                    # Maintain a dict of device names
+                    device = self._sanitize_dbus_value(udisk_block["Device"])
+                    self._drive_by_device_name[udisk_block["Drive"]] = device
+
+            except Exception as ae:
+                print("block_dev unusable: %r" % ae)
+
+    def _sanitize_dbus_value(self, value):
+        """Convert certain DBus type combinations so that they are easier to read"""
+        if isinstance(value, Array) and value.signature == "ay":
+            try:
+                return self._decode_ay(value)
+            except:
+                # Try an array of arrays; 'aay' which is the symlinks
+                return list(map(self._decode_ay, value or ()))
+        elif isinstance(value, Array) and value.signature == "y":
+            return bytearray(value).rstrip(bytearray((0,))).decode('utf-8')
+        else:
+            return value
+
+    def _decode_ay(self, value):
+        """Convert binary blob from DBus queries to strings"""
+        if len(value) == 0 or \
+           value is None:
+            return ''
+        elif isinstance(value, str):
+            return value
+        elif isinstance(value, bytes):
+            return value.decode('utf-8')
+        else:
+            # dbus.Array([dbus.Byte]) or any similar sequence type:
+            return bytearray(value).rstrip(bytearray((0,))).decode('utf-8')
+
 
+    def get_psu_info(self):
+        """Update and return PSU information in specific format."""
+        psus = []
+        response, _, _ = SimpleProcess("dmidecode").run()
```

Comment:
Can limit to psu type

>]# dmidecode -t 39
# dmidecode 3.2
Getting SMBIOS data from sysfs.
SMBIOS 3.2.1 present.
# SMBIOS implementations newer than version 3.2.0 are not
# fully supported by this version of dmidecode.

Handle 0x000C, DMI type 39, 22 bytes
System Power Supply
        Power Unit Group: 1
        Location: PSU1
        Name: PWS-751P-1R
        Manufacturer: SUPERMICRO
        Serial Number: P751PCJ31CU1144
        Asset Tag: N/A
        Model Part Number: PWS-751P-1R
        Revision: 1.1
        Max Power Capacity: 750 W
        Status: Present, OK
        Type: Switching
        Input Voltage Range Switching: Auto-switch
        Plugged: Yes
        Hot Replaceable: Yes

Handle 0x000D, DMI type 39, 22 bytes
System Power Supply
        Power Unit Group: 2
        Location: PSU2
        Name: PWS-751P-1R
        Manufacturer: SUPERMICRO
        Serial Number: P751PCJ31CU1107
        Asset Tag: N/A
        Model Part Number: PWS-751P-1R
        Revision: 1.1
        Max Power Capacity: 750 W
        Status: Present, OK
        Type: Switching
        Input Voltage Range Switching: Auto-switch
        Plugged: Yes
        Hot Replaceable: Yes